v2-committee: improve concurrency handling

iurii-ssv opened the pull request:
This PR intends to:
- reduce mutex contention by moving potentially long-lasting call (which contains network request I believe) out of section-under-lock
- fix potentially dangerous mutex usage in `HandleMessage` (due to the way it currently locks mutex, it potentially can send message into queue that will get garbage-collected)
- clean up, standardize and clarify things

See additional comments below.

iurii-ssv commented on protocol/v2/ssv/validator/committee.go:
```diff
@@ -90,120 +89,89 @@ func (c *Committee) RemoveShare(validatorIndex phase0.ValidatorIndex) {
 	return nil
 }
 
-func (c *Committee) PushToQueue(slot phase0.Slot, dec *queue.SSVMessage) {
-	c.mtx.RLock()
-	queue, exists := c.Queues[slot]
-	c.mtx.RUnlock()
-	if !exists {
-		c.logger.Warn("cannot push to non-existing queue", zap.Uint64("slot", uint64(slot)))
-		return
-	}
-	if pushed := queue.Q.TryPush(dec); !pushed {
-		c.logger.Warn("dropping ExecuteDuty message because the queue is full")
-	}
-}
```
Unused method, cleaning up.

iurii-ssv commented on protocol/v2/ssv/validator/committee_queue.go:
```diff
@@ -30,24 +30,28 @@ func (c *Committee) HandleMessage(logger *zap.Logger, msg *queue.SSVMessage) {
+
+		return q
 	}
+	q := prepareQueue()
```
Here we potentially can miss a message if `HandleMessage` is called concurrently 2 or more times(if we just lock `c.mtx` for read and then potentially add new Queue - concurrent caller might do the same), 

not sure if this is a real problem in practice, just seems to me it's better not to make additional assumptions and do this lazy initialization in safe manner (which is what `prepareQueue` does).

iurii-ssv commented on protocol/v2/ssv/validator/committee_queue.go:
```diff
@@ -59,26 +63,39 @@ func (c *Committee) HandleMessage(logger *zap.Logger, msg *queue.SSVMessage) {
 	}
 }
 
-//// StartQueueConsumer start ConsumeQueue with handler
-//func (v *Committee) StartQueueConsumer(logger *zap.Logger, msgID spectypes.MessageID, handler MessageHandler) {
-//	ctx, cancel := context.WithCancel(v.ctx)
-//	defer cancel()
-//
-//	for ctx.Err() == nil {
-//		err := v.ConsumeQueue(logger, msgID, handler)
-//		if err != nil {
-//			logger.Debug("❗ failed consuming queue", zap.Error(err))
-//		}
-//	}
-//}
```
Dead code cleanup.

nkryuchkov commented on protocol/v2/ssv/validator/committee.go:
```diff
@@ -231,13 +199,13 @@ func (c *Committee) ProcessMessage(logger *zap.Logger, msg *queue.SSVMessage) er
+		c.mtx.RUnlock()
 		if !exists {
-			return errors.New("no runner found for message's slot")
+			return errors.New(fmt.Sprintf("no runner found for consensus message's slot(qbtf height): %d", qbftMsg.Height))
```
We always use slot as QBFT height, so we don't have to clarify this in logs
```suggestion
			return fmt.Errorf("no runner found for consensus message's slot %d", qbftMsg.Height)
```

iurii-ssv replied:
Applied in https://github.com/ssvlabs/ssv/pull/1784/commits/483c2ed41d9b789251d57e9d6ce4f3275e3f02c1

nkryuchkov commented on protocol/v2/ssv/validator/committee.go:
```diff
@@ -90,120 +89,89 @@ func (c *Committee) RemoveShare(validatorIndex phase0.ValidatorIndex) {
 func (c *Committee) StartDuty(logger *zap.Logger, duty *spectypes.CommitteeDuty) error {
-	c.mtx.Lock()
-	defer c.mtx.Unlock()
+	prepareRunner := func() (runner.Runner, error) {
```
Since we touch it, can we extract it into another method for readability? Perhaps we could split this as well

iurii-ssv replied:
Yep, addressed in https://github.com/ssvlabs/ssv/pull/1784/commits/483c2ed41d9b789251d57e9d6ce4f3275e3f02c1

nkryuchkov commented on protocol/v2/ssv/validator/committee.go:
```diff
@@ -254,22 +222,22 @@ func (c *Committee) ProcessMessage(logger *zap.Logger, msg *queue.SSVMessage) er
+			c.mtx.RUnlock()
 			if !exists {
-				return errors.New("no runner found for message's slot")
+				return errors.New(fmt.Sprintf("no runner found for post consensus sig message's slot: %d", pSigMessages.Slot))
```
```suggestion
				return fmt.Errorf("no runner found for post consensus sig message's slot %d", pSigMessages.Slot)
```

iurii-ssv replied:
Applied in https://github.com/ssvlabs/ssv/pull/1784/commits/483c2ed41d9b789251d57e9d6ce4f3275e3f02c1

nkryuchkov commented on protocol/v2/ssv/validator/committee_queue.go:
```diff
@@ -30,24 +30,28 @@ func (c *Committee) HandleMessage(logger *zap.Logger, msg *queue.SSVMessage) {
-				//Quorum:             options.SSVShare.Share,// TODO
-			},
-		}
+	prepareQueue := func() queueContainer {
```
can we extract this into a new method for readability?

iurii-ssv replied:
Yep, addressed in https://github.com/ssvlabs/ssv/pull/1784/commits/483c2ed41d9b789251d57e9d6ce4f3275e3f02c1

nkryuchkov reviewed (commented):
LGTM, but I'd try to make just a few minor further simplifications as we're touching this

nkryuchkov reviewed (approved):
LGTM, let's just make sure all CI checks pass

nkryuchkov commented on protocol/v2/ssv/validator/committee.go:
```diff
@@ -90,80 +89,43 @@ func (c *Committee) RemoveShare(validatorIndex phase0.ValidatorIndex) {
-		return errors.New("no shares for duty's validators")
+	shares, attesters, trimmedDuty, err := c.prepareDuty(logger, duty)
+	if err != nil {
+		return nil, fmt.Errorf("prepareDutyRunner: %w", err)
```
```suggestion
		return nil, fmt.Errorf("prepare duty: %w", err)
```

or 

```suggestion
		return nil, fmt.Errorf("prepareDuty: %w", err)
```

iurii-ssv commented on scripts/spec-alignment/differ.config.yaml:
```diff
@@ -1,4 +1,4 @@
-ApprovedChanges: ["50e5bb7eda99594e", "870a3a66aeccd737","4e22a08543b079b","56ceb03cd44ff702","188adfe8914e04c1","2438f9c5b82b69a3","1a716ee3bdb3170","90b166f78390af18","68219b82a1d9d829","c4c4caa5d0938b85","dfe99ce1d27b6cb1","35f5dab1f128d193","9a3973b64d7e8932","f33f07301a770d03","3e9e0dddfad3b302","d4fef6512374c1f5","b49f54cb45787e4b","59b2375130aef5df","f094cd0460432170","8e51881e527dd603","a ...
+ApprovedChanges: ["50e5bb7eda99594e", "870a3a66aeccd737","4e22a08543b079b","56ceb03cd44ff702","188adfe8914e04c1","2438f9c5b82b69a3","1a716ee3bdb3170","90b166f78390af18","68219b82a1d9d829","c4c4caa5d0938b85","dfe99ce1d27b6cb1","35f5dab1f128d193","9a3973b64d7e8932","f33f07301a770d03","3e9e0dddfad3b302","d4fef6512374c1f5","b49f54cb45787e4b","59b2375130aef5df","f094cd0460432170","8e51881e527dd603","a ...
```
I assume this is what we do to satisfy `spec-alignment` pipeline, the diff generated for this PR looks like this:
```
diff --git a/protocol/v2/ssv/validator/committee.go@Committee b/ssv/committee.go@Committee
index b40d693202be4382..2222222
--- a/protocol/v2/ssv/validator/committee.go@Committee
+++ b/ssv/committee.go@Committee
@@ -1,14 +1,6 @@
 Committee struct {
-	logger	*zap.Logger
-	ctx	context.Context
-	cancel	context.CancelFunc
-	BeaconNetwork	BeaconNetwork
-	Storage		*storage.QBFTStores
-	mtx	sync.RWMutex
-	Queues	map[phase0.Slot]queueContainer
-	Runners	map[phase0.Slot]*runner.CommitteeRunner
-	Shares	map[phase0.ValidatorIndex]*Share
-	CommitteeMember	*CommitteeMember
-	dutyGuard	*CommitteeDutyGuard
-	CreateRunnerFn	CommitteeRunnerFunc
+	Runners		map[spec.Slot]*CommitteeRunner
+	CommitteeMember	CommitteeMember
+	CreateRunnerFn	CreateRunnerFn
+	Share		map[spec.ValidatorIndex]*Share
 }
\ No newline at end of file
diff --git a/protocol/v2/ssv/validator/committee.go@Committee.ProcessMessage b/ssv/committee.go@Committee.ProcessMessage
index 9129f80a9f1b1331..2222222
--- a/protocol/v2/ssv/validator/committee.go@Committee.ProcessMessage
+++ b/ssv/committee.go@Committee.ProcessMessage
@@ -1,55 +1,48 @@
-func (c *Committee) ProcessMessage(msg *queue.SSVMessage) error {
-	if msg.GetType() != message.SSVEventMsgType {
-		if err := msg.SignedSSVMessage.Validate(); err != nil {
-			return errors.Wrap(err, "invalid SignedSSVMessage")
-		}
-		if err := Verify(msg.SignedSSVMessage, c.CommitteeMember.Committee); err != nil {
-			return errors.Wrap(err, "SignedSSVMessage has an invalid signature")
-		}
-		if err := c.validateMessage(msg.SignedSSVMessage.SSVMessage); err != nil {
-			return errors.Wrap(err, "Message invalid")
-		}
+func (c *Committee) ProcessMessage(signedSSVMessage *SignedSSVMessage) error {
+	if err := signedSSVMessage.Validate(); err != nil {
+		return errors.Wrap(err, "invalid SignedSSVMessage")
+	}
+	if err := Verify(signedSSVMessage, c.CommitteeMember.Committee); err != nil {
+		return errors.Wrap(err, "SignedSSVMessage has an invalid signature")
+	}
+	msg := signedSSVMessage.SSVMessage
+	if err := c.validateMessage(msg); err != nil {
+		return errors.Wrap(err, "Message invalid")
 	}
 	switch msg.GetType() {
 	case SSVConsensusMsgType:
 		qbftMsg := &Message{}
 		if err := qbftMsg.Decode(msg.GetData()); err != nil {
 			return errors.Wrap(err, "could not get consensus Message from network Message")
 		}
 		if err := qbftMsg.Validate(); err != nil {
 			return errors.Wrap(err, "invalid qbft Message")
 		}
-		c.mtx.RLock()
-		r, exists := c.Runners[phase0.Slot(qbftMsg.Height)]
-		c.mtx.RUnlock()
+		runner, exists := c.Runners[spec.Slot(qbftMsg.Height)]
 		if !exists {
-			return fmt.Errorf("no runner found for consensus message's slot %d", qbftMsg.Height)
+			return errors.New("no runner found for message's slot")
 		}
-		return r.ProcessConsensus(msg.SignedSSVMessage)
+		return runner.ProcessConsensus(signedSSVMessage)
 	case SSVPartialSignatureMsgType:
 		pSigMessages := &PartialSignatureMessages{}
-		if err := pSigMessages.Decode(msg.SignedSSVMessage.SSVMessage.GetData()); err != nil {
+		if err := pSigMessages.Decode(msg.GetData()); err != nil {
 			return errors.Wrap(err, "could not get post consensus Message from network Message")
 		}
-		if len(msg.SignedSSVMessage.OperatorIDs) != 1 {
+		if len(signedSSVMessage.OperatorIDs) != 1 {
 			return errors.New("PartialSignatureMessage has more than 1 signer")
 		}
-		if err := pSigMessages.ValidateForSigner(msg.SignedSSVMessage.OperatorIDs[0]); err != nil {
+		if err := pSigMessages.ValidateForSigner(signedSSVMessage.OperatorIDs[0]); err != nil {
 			return errors.Wrap(err, "invalid PartialSignatureMessages")
 		}
 		if pSigMessages.Type == PostConsensusPartialSig {
-			c.mtx.RLock()
-			r, exists := c.Runners[pSigMessages.Slot]
-			c.mtx.RUnlock()
+			runner, exists := c.Runners[pSigMessages.Slot]
 			if !exists {
-				return fmt.Errorf("no runner found for post consensus sig message's slot %d", pSigMessages.Slot)
+				return errors.New("no runner found for message's slot")
 			}
-			return r.ProcessPostConsensus(pSigMessages)
+			return runner.ProcessPostConsensus(pSigMessages)
 		}
-	case message.SSVEventMsgType:
-		return c.handleEventMessage(msg)
 	default:
 		return errors.New("unknown msg")
 	}
 	return nil
 }
\ No newline at end of file
diff --git a/protocol/v2/ssv/validator/committee.go@Committee.StartDuty b/ssv/committee.go@Committee.StartDuty
index b3b73f9cdf1c39fb..2222222
--- a/protocol/v2/ssv/validator/committee.go@Committee.StartDuty
+++ b/ssv/committee.go@Committee.StartDuty
@@ -1,11 +1,24 @@
 func (c *Committee) StartDuty(duty *CommitteeDuty) error {
-	r, trimmedDuty, err := c.prepareDutyRunner(duty)
-	if err != nil {
-		return fmt.Errorf("could not prepare duty runner: %w", err)
+	if len(duty.ValidatorDuties) == 0 {
+		return errors.New("no beacon duties")
 	}
-	err = r.StartNewDuty(trimmedDuty, c.CommitteeMember.GetQuorum())
-	if err != nil {
-		return errors.Wrap(err, "runner failed to start duty")
+	if _, exists := c.Runners[duty.Slot]; exists {
+		return errors.New(fmt.Sprintf("CommitteeRunner for slot %d already exists", duty.Slot))
 	}
-	return nil
+	dutyShares := make(map[spec.ValidatorIndex]*Share)
+	filteredDuty := &CommitteeDuty{
+		Slot: duty.Slot,
+	}
+	for _, bduty := range duty.ValidatorDuties {
+		if _, exists := c.Share[bduty.ValidatorIndex]; !exists {
+			continue
+		}
+		dutyShares[bduty.ValidatorIndex] = c.Share[bduty.ValidatorIndex]
+		filteredDuty.ValidatorDuties = append(filteredDuty.ValidatorDuties, bduty)
+	}
+	if len(dutyShares) == 0 {
+		return errors.New("no shares for duty's validators")
+	}
+	c.Runners[filteredDuty.Slot] = c.CreateRunnerFn(dutyShares)
+	return c.Runners[filteredDuty.Slot].StartNewDuty(filteredDuty, c.CommitteeMember.GetQuorum())
 }
\ No newline at end of file
```

Not sure why it works like this - but doesn't feel very useful to have `spec-alignment` fail when implementation detail change. Perhaps we can change it somehow to only call out `func` / `interface` signature updates ?

iurii-ssv replied:
Updated spec alignment (after rebase) - https://github.com/ssvlabs/ssv/pull/1784/commits/7e529a404b07d724402c44b188dc44baa99329ac, new diff:
```
diff --git a/protocol/v2/ssv/validator/committee.go@Committee b/ssv/committee.go@Committee
index 4418aa6010632480..2222222
--- a/protocol/v2/ssv/validator/committee.go@Committee
+++ b/ssv/committee.go@Committee
@@ -1,13 +1,6 @@
 Committee struct {
-	logger	*zap.Logger
-	ctx	context.Context
-	cancel	context.CancelFunc
-	BeaconNetwork	BeaconNetwork
-	mtx	sync.RWMutex
-	Queues	map[phase0.Slot]queueContainer
-	Runners	map[phase0.Slot]*runner.CommitteeRunner
-	Shares	map[phase0.ValidatorIndex]*Share
-	CommitteeMember	*CommitteeMember
-	dutyGuard	*CommitteeDutyGuard
-	CreateRunnerFn	CommitteeRunnerFunc
+	Runners		map[spec.Slot]*CommitteeRunner
+	CommitteeMember	CommitteeMember
+	CreateRunnerFn	CreateRunnerFn
+	Share		map[spec.ValidatorIndex]*Share
 }
\ No newline at end of file
diff --git a/protocol/v2/ssv/validator/committee.go@Committee.ProcessMessage b/ssv/committee.go@Committee.ProcessMessage
index 4cf190dd314b064b..2222222
--- a/protocol/v2/ssv/validator/committee.go@Committee.ProcessMessage
+++ b/ssv/committee.go@Committee.ProcessMessage
@@ -1,55 +1,48 @@
-func (c *Committee) ProcessMessage(msg *queue.SSVMessage) error {
-	if msg.GetType() != message.SSVEventMsgType {
-		if err := msg.SignedSSVMessage.Validate(); err != nil {
-			return errors.Wrap(err, "invalid SignedSSVMessage")
-		}
-		if err := Verify(msg.SignedSSVMessage, c.CommitteeMember.Committee); err != nil {
-			return errors.Wrap(err, "SignedSSVMessage has an invalid signature")
-		}
-		if err := c.validateMessage(msg.SignedSSVMessage.SSVMessage); err != nil {
-			return errors.Wrap(err, "Message invalid")
-		}
+func (c *Committee) ProcessMessage(signedSSVMessage *SignedSSVMessage) error {
+	if err := signedSSVMessage.Validate(); err != nil {
+		return errors.Wrap(err, "invalid SignedSSVMessage")
+	}
+	if err := Verify(signedSSVMessage, c.CommitteeMember.Committee); err != nil {
+		return errors.Wrap(err, "SignedSSVMessage has an invalid signature")
+	}
+	msg := signedSSVMessage.SSVMessage
+	if err := c.validateMessage(msg); err != nil {
+		return errors.Wrap(err, "Message invalid")
 	}
 	switch msg.GetType() {
 	case SSVConsensusMsgType:
 		qbftMsg := &Message{}
 		if err := qbftMsg.Decode(msg.GetData()); err != nil {
 			return errors.Wrap(err, "could not get consensus Message from network Message")
 		}
 		if err := qbftMsg.Validate(); err != nil {
 			return errors.Wrap(err, "invalid qbft Message")
 		}
-		c.mtx.RLock()
-		r, exists := c.Runners[phase0.Slot(qbftMsg.Height)]
-		c.mtx.RUnlock()
+		runner, exists := c.Runners[spec.Slot(qbftMsg.Height)]
 		if !exists {
-			return fmt.Errorf("no runner found for consensus message's slot %d", qbftMsg.Height)
+			return errors.New("no runner found for message's slot")
 		}
-		return r.ProcessConsensus(msg.SignedSSVMessage)
+		return runner.ProcessConsensus(signedSSVMessage)
 	case SSVPartialSignatureMsgType:
 		pSigMessages := &PartialSignatureMessages{}
-		if err := pSigMessages.Decode(msg.SignedSSVMessage.SSVMessage.GetData()); err != nil {
+		if err := pSigMessages.Decode(msg.GetData()); err != nil {
 			return errors.Wrap(err, "could not get post consensus Message from network Message")
 		}
-		if len(msg.SignedSSVMessage.OperatorIDs) != 1 {
+		if len(signedSSVMessage.OperatorIDs) != 1 {
 			return errors.New("PartialSignatureMessage has more than 1 signer")
 		}
-		if err := pSigMessages.ValidateForSigner(msg.SignedSSVMessage.OperatorIDs[0]); err != nil {
+		if err := pSigMessages.ValidateForSigner(signedSSVMessage.OperatorIDs[0]); err != nil {
 			return errors.Wrap(err, "invalid PartialSignatureMessages")
 		}
 		if pSigMessages.Type == PostConsensusPartialSig {
-			c.mtx.RLock()
-			r, exists := c.Runners[pSigMessages.Slot]
-			c.mtx.RUnlock()
+			runner, exists := c.Runners[pSigMessages.Slot]
 			if !exists {
-				return fmt.Errorf("no runner found for post consensus partial-sig message's slot %d", pSigMessages.Slot)
+				return errors.New("no runner found for message's slot")
 			}
-			return r.ProcessPostConsensus(pSigMessages)
+			return runner.ProcessPostConsensus(pSigMessages)
 		}
-	case message.SSVEventMsgType:
-		return c.handleEventMessage(msg)
 	default:
 		return errors.New("unknown msg")
 	}
 	return nil
 }
\ No newline at end of file
diff --git a/protocol/v2/ssv/validator/committee.go@Committee.StartDuty b/ssv/committee.go@Committee.StartDuty
index b3b73f9cdf1c39fb..2222222
--- a/protocol/v2/ssv/validator/committee.go@Committee.StartDuty
+++ b/ssv/committee.go@Committee.StartDuty
@@ -1,11 +1,24 @@
 func (c *Committee) StartDuty(duty *CommitteeDuty) error {
-	r, trimmedDuty, err := c.prepareDutyRunner(duty)
-	if err != nil {
-		return fmt.Errorf("could not prepare duty runner: %w", err)
+	if len(duty.ValidatorDuties) == 0 {
+		return errors.New("no beacon duties")
 	}
-	err = r.StartNewDuty(trimmedDuty, c.CommitteeMember.GetQuorum())
-	if err != nil {
-		return errors.Wrap(err, "runner failed to start duty")
+	if _, exists := c.Runners[duty.Slot]; exists {
+		return errors.New(fmt.Sprintf("CommitteeRunner for slot %d already exists", duty.Slot))
 	}
-	return nil
+	dutyShares := make(map[spec.ValidatorIndex]*Share)
+	filteredDuty := &CommitteeDuty{
+		Slot: duty.Slot,
+	}
+	for _, bduty := range duty.ValidatorDuties {
+		if _, exists := c.Share[bduty.ValidatorIndex]; !exists {
+			continue
+		}
+		dutyShares[bduty.ValidatorIndex] = c.Share[bduty.ValidatorIndex]
+		filteredDuty.ValidatorDuties = append(filteredDuty.ValidatorDuties, bduty)
+	}
+	if len(dutyShares) == 0 {
+		return errors.New("no shares for duty's validators")
+	}
+	c.Runners[filteredDuty.Slot] = c.CreateRunnerFn(dutyShares)
+	return c.Runners[filteredDuty.Slot].StartNewDuty(filteredDuty, c.CommitteeMember.GetQuorum())
 }
\ No newline at end of file
```

iurii-ssv commented on protocol/v2/ssv/spectest/committee_msg_processing_type.go:
```diff
@@ -45,7 +44,9 @@ func (test *CommitteeSpecTest) RunAsPartOfMultiTest(t *testing.T) {
 	lastErr := test.runPreTesting(logger)
 
 	if len(test.ExpectedError) != 0 {
-		require.EqualError(t, lastErr, test.ExpectedError)
+		// Don't want to compare error messages exactly because it's an implementation detail that
+		// might differ from the spec.
+		require.Error(t, lastErr)
 	} else {
 		require.NoError(t, lastErr)
 	}
```
Seems like we **should be able to update error messages** (to add additional info in there, and make it more detailed if necessary) even if they diverge from those we have in spec ?

Could update spec messages instead, but seems like it's an implementation detail not worth spending time on, WDYT ?

iurii-ssv replied:
Actually even better approach here would probably be:
- to define a mapping (here in `ssv` repo) `testName`->`ExpectedErrorOverride`
- leave the check as it was (exact error string match) but take `ExpectedErrorOverride` value (instead of `ExpectedError`) if it exists in that mapping we added
- and then, for each error-message update like I have in this PR, we just have to manually verify each failing test (expected vs got error message) & override those errors where message differs now; checking for exact string match against `ExpectedErrorOverride` wont work though because error messages will have dynamic data in them (such as expected vs actual data and such), so regex-based matching should do the job

iurii-ssv replied:
@y0sher, @moshe-blox would be nice to get an ACK on this from one of you guys, and I think this PR is ready to merge

iurii-ssv commented:
This PR seems to be working fine on stage.

y0sher reviewed (approved):
Looks good in terms of code. The implementation of `Committee` was intentionally simple and "under-optimized" to increase development speed. This is a good time to fine tune these stuff.
please lets test this code and see what impact it has.

iurii-ssv commented:
> please lets test this code and see what impact it has.

Running it on **stage** for past couple of hours, doesn't seem to be worse (compared to how `stage` branch runs):

<img width="600" alt="image" src="https://github.com/user-attachments/assets/15fd64be-d2d2-45f0-906f-1dbb2ae833e1">

![image](https://github.com/user-attachments/assets/a1d2913f-8d77-4834-88d7-d0d27dd0f294)

And spec-tests actually did discover a bug that I fixed today https://github.com/ssvlabs/ssv/pull/1784/commits/d3756d0e89d8e5e35adadcf78ffe95c8a5b73ac0, and now pipelines for this PR are passing - I pushed [this branch](https://github.com/ssvlabs/ssv/tree/v2-committee-improve-concurrency-handling) to [ssv repo ](https://github.com/ssvlabs/ssv) mirroring these changes to trigger CI for this PR.

Or did you mean we'd want more unit-tests for this functionality **before** "it's safe to merge" ?

iurii-ssv commented:
Just rebased, CI is passing - https://github.com/ssvlabs/ssv/actions/runs/11636222622, the only notable change I've added since last review(s) is https://github.com/ssvlabs/ssv/pull/1784#discussion_r1818179653
